src/osp/Active/SysNewton.cpp

```diff
@@ -94,66 +99,129 @@ ACompNwtBody& ACompNwtBody::operator=(ACompNwtBody&& move)
+ : m_scene(scene)
+ , m_updatePhysicsWorld(scene.get_update_order(), "physics", "wire", "",
+                        [this] { this->update_world(); })
 {
     //std::cout << "sysnewtoninit\n";
-    NewtonWorldSetUserData(m_nwtWorld, this);
+    //NewtonWorldSetUserData(m_nwtWorld, this);
+
+    // Connect signal handlers to destruct Newton objects when components are
+    // deleted.
 
-    scene.get_registry().on_construct<ACompRigidBody>()
-                    .connect<&SysNewton::on_body_construct>(*this);
-    scene.get_registry().on_destroy<ACompRigidBody>()
-                    .connect<&SysNewton::on_body_destruct>(*this);
+    scene.get_registry().on_destroy<ACompNwtBody>()
+                    .connect<&SysNewton::on_body_destruct>();
 
-    scene.get_registry().on_construct<ACompCollisionShape>()
-                    .connect<&SysNewton::on_shape_construct>(*this);
     scene.get_registry().on_destroy<ACompCollisionShape>()
-                    .connect<&SysNewton::on_shape_destruct>(*this);
+                    .connect<&SysNewton::on_shape_destruct>();
+
+    scene.get_registry().on_destroy<ACompNwtWorld>()
+                    .connect<&SysNewton::on_world_destruct>();
 }
 
 SysNewton::~SysNewton()
 {
     //std::cout << "sysnewtondestroy\n";
     // Clean up newton dynamics stuff
-    m_scene.get_registry().clear<ACompRigidBody>();
+    m_scene.get_registry().clear<ACompNwtBody>();
     m_scene.get_registry().clear<ACompCollisionShape>();
-    NewtonDestroyAllBodies(m_nwtWorld);
-    NewtonDestroy(m_nwtWorld);
+    //NewtonDestroyAllBodies(m_nwtWorld);
+    //NewtonDestroy(m_nwtWorld);
 }
 
-void SysNewton::find_and_add_colliders(ActiveEnt ent, NewtonCollision *compound,
-                                       Matrix4 const &currentTransform)
+void SysNewton::update_world()
 {
+    // Try getting the newton world from the scene root.
+    // It is very unlikely that multiple physics worlds are ever needed
+    auto *worldComp = m_scene.get_registry().try_get<ACompNwtWorld>(
+                m_scene.hier_get_root());
+
+    if (worldComp == nullptr)
+    {
+        return; // No physics world component
+    }
+
+    if (worldComp->m_nwtWorld == nullptr)
+    {
+        // Create world if not yet initialized
+        worldComp->m_nwtWorld = NewtonCreate();
+        NewtonWorldSetUserData(worldComp->m_nwtWorld, &m_scene);
+    }
+
+    NewtonWorld const* nwtWorld = worldComp->m_nwtWorld;
+
+    auto viewBody = m_scene.get_registry().view<ACompNwtBody>();
+    auto viewBodyTransform = m_scene.get_registry().view<ACompNwtBody,
+                                                         ACompTransform>();
+
+    // Iterate just the entities with ACompNwtBody
+    for (ActiveEnt ent : viewBody)
+    {
+        auto &entBody = viewBodyTransform.get<ACompNwtBody>(ent);
+
+        // temporary: delete if something is dirty
+        if (entBody.m_colliderDirty)
+        {
+            NewtonDestroyBody(entBody.m_body);
```

Comment:
A long time ago I learned a very painful mistake about setting things to nullptr *after* destroying them. The long and short of it is that someone's gonna write some code that, somehow, manages to access your just-destroyed thing *after* it was destroyed but before it was set to nullptr, which then makes the sr. director of a different department call your boss really angerly demanding that you fix the code you broke, even though it turns out it was their code that was broken the whole time but it only managed to work by accident because the operating system didn't clear out the pointed-to-object after destruction so accessing it after it was freed still magically worked until i fixed my code.

Reply:
As such, i recommend

NewtonDestroyBody(std::exchange(entBody.m_body, nullptr));

Reply:
Hmm, interesting disaster. I hope the accepted fix was not removing the `= nullptr`

Reply:
Na, we fixed it properly.

Basically
```
auto pOld = std::exchange(m_p, nullptr);
delete pOld;
return;
```
instead of
```
delete m_p;
m_p = nullptr;
return;
```